components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java

```diff
@@ -259,23 +260,32 @@ private boolean revokeTokens(String username, UserStoreManager userStoreManager)
                 log.error(errorMsg, e);
                 return true;
             }
+
+            Set<String> scopes = new HashSet<>();
             for (AccessTokenDO accessTokenDO : accessTokenDOs) {
-                //Clear cache
+                // Clear cache
                 OAuthUtil.clearOAuthCache(accessTokenDO.getConsumerKey(), accessTokenDO.getAuthzUser(),
                         OAuth2Util.buildScopeString(accessTokenDO.getScope()));
                 OAuthUtil.clearOAuthCache(accessTokenDO.getConsumerKey(), accessTokenDO.getAuthzUser());
                 OAuthUtil.clearOAuthCache(accessTokenDO.getAccessToken());
+
+                // Get unique scopes list
+                scopes.add(OAuth2Util.buildScopeString(accessTokenDO.getScope()));
```

Comment:
If there are same set of scopes for multiple access tokens, HashSet won't allow multiple entries with same value. So seems like, there can be situation that all the ACTIVE and EXPIRED tokens will not get REVOKED from database (If there are multiple access tokens with same scopes, only the latest access token will moved to REVOKED state). WDYT?

Reply:
I guess given user, given scope, given client id only one access token exist

Reply:
For given user, given scope, given client id, only one ACTIVE token exists, there can be EXPIRED tokens. Previously we move all the ACTIVE and EXPIRED access tokens to REVOKED state [1]. So with this fix, it'll be changed to, move only the latest (ACTIVE or EXPIRED) access token into REVOKED state, for tokens with same scopes.

[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOathEventListener.java#L254

Reply:
good catch! but I don't see a major impact. Let's have a discussion

Reply:
@IndunilRathnayake This has not changed the previous behaviour IMO. As for the previous implementation also we only revoke the latest access token for a given **user, client id, scope** combination as we are using the getLatestAccessToken method as [here](https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/822/files#diff-0daa59162506998f3129964e2307f192R282).